File: ML in Python & R/Part 2 - Regression/Section 5 - Multiple Linear Regression/MyWork/multiple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 24 21:07:22 2020

@author: chath
"""

## Data Preprocessing

# Import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Import dataset
dataset = pd.read_csv('50_Startups.csv')

x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

## Encoding categorical data
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

columnTransformer = ColumnTransformer(
    transformers = [('one_hot_encoder', OneHotEncoder(categories='auto'), [3])],
    remainder = 'passthrough')
x = columnTransformer.fit_transform(x)

## Avoiding the dummy variable trap
x = x[:, 1:] # we remove one of categorical dummy variable to prevent dummy variable trap

## Split dataset to train and test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

## Feature scaling 
"""
# There are two ways Standardisation and Normalisation
from sklearn.preprocessing import StandardScaler

sc_x = StandardScaler()
# For the train data set we need to fit and transaform
x_train = sc_x.fit_transform(x_train)
# sc_x already fitted to training set, so just need transaform
x_test = sc_x.transform(x_test)
"""

## Fitting multiple linear regression to the training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(x_train, y_train)

## Predicting the test set results
y_pred = regressor.predict(x_test)

## Building the optimal model using Backward elimination
import statsmodels.api as sm
# Add coumns of ones
x = np.append(arr = np.ones((50, 1)).astype(int), values = x, axis = 1)
# Create optimal matris of features- only contain highly impacting variables
x_opt = x[:, [0, 1, 2, 3, 4, 5]]
x_opt = np.array(x_opt, dtype = float)
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

# Let's take significate level as 0.05 which is 5%
# In backward elimination remove first hight P valued parameter
# x2 has 0.990 which is 99%, which greater than 5% so it need to remove
x_opt = x[:, [0, 1, 3, 4, 5]]
x_opt = np.array(x_opt, dtype = float)
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

# x1 has 0.940 which is 94%, which greater than 5% so it need to remove
x_opt = x[:, [0, 3, 4, 5]]
x_opt = np.array(x_opt, dtype = float)
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

# x2 has 0.602 which is 60.2%, which greater than 5% so it need to remove
x_opt = x[:, [0, 3, 5]]
x_opt = np.array(x_opt, dtype = float)
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

# x2 has 0.060 which is 6%, which greater than 5% so it need to remove
x_opt = x[:, [0, 3]]
x_opt = np.array(x_opt, dtype = float)
regressor_OLS = sm.OLS(endog = y, exog = x_opt).fit()
regressor_OLS.summary()

File: ML in Python & R/Part 2 - Regression/Section 5 - Multiple Linear Regression/MyWork/multiple_linear_regression.R

### Data preprocessing

## Import dataset
dataset = read.csv('50_Startups.csv')
# dataset = dataset[, 2:3]

## Encoding categorical data
# R see Categories as Facors so no need three columns

dataset$State = factor(x = dataset$State, 
                         levels = c('New York', 'California', 'Florida'),
                         labels = c(1, 2, 3))

## Split dataset to train and test

# Need a new package caTools
# install.packages('caTools')

library(caTools) #using this we can import caTools directly

# In python random_state, in R it is seed
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8) # SplitRatio is train set ratio oposite of python, This returns True - decide to goto train/ False- goto test set

training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)


# ## Feature scaling
# # Facot is not numeric
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

## Fitting Muliple Linear Regression to training set
regressor = lm(formula = Profit ~ .,
               data = training_set) # Instead of Profit ~ R.D.Spend + Administration + Marketing.Spend + State we can use Profit ~ . <- means all
 
## Predicting the test set result
y_pred = predict(regressor, newdata = test_set)

## Building optimal model using Backward Elimination

# State 2, State 3 means dummy variables. R manages it.
# Summary(resressor) in console to get summary of regressor
# look at significat (P Value) take significate

# Fitting Muliple Linear Regression to training set with indipendant variables
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = training_set)
summary(regressor)

# State 2 has highest P-value
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = training_set)
summary(regressor)

# Administration has highest P-value
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = training_set)
summary(regressor)

# Remove  Marketing.Spend
regressor = lm(formula = Profit ~ R.D.Spend,
               data = training_set)
summary(regressor)













